Realm/RLMManagedSet.mm

```diff
@@ -0,0 +1,564 @@
+
+    translateErrors(f);
+}
+
+//
+// public method implementations
+//
+- (RLMRealm *)realm {
+    return _realm;
+}
+
+- (NSUInteger)count {
+    return translateErrors([&] { return _backingSet.size(); });
+}
+
+- (NSArray<id> *)allObjects {
+    NSMutableArray *arr = [NSMutableArray new];
+    for (id prop : self) {
+        [arr addObject:prop];
+    }
+    return arr;
+}
+
+- (BOOL)isInvalidated {
+    return translateErrors([&] { return !_backingSet.is_valid(); });
+}
+
+- (RLMClassInfo *)objectInfo {
+    return _objectInfo;
+}
+
+
+- (bool)isBackedBySet:(realm::object_store::Set const&)set {
+    return _backingSet == set;
+}
+
+- (BOOL)isEqual:(id)object {
+    return [object respondsToSelector:@selector(isBackedBySet:)] && [object isBackedBySet:_backingSet];
+}
+
+- (NSUInteger)hash {
+    return std::hash<realm::object_store::Set>()(_backingSet);
+}
+
+- (NSUInteger)countByEnumeratingWithState:(NSFastEnumerationState *)state
+                                  objects:(__unused __unsafe_unretained id [])buffer
+                                    count:(NSUInteger)len {
+    return RLMFastEnumerate(state, len, self);
+}
+
+static void RLMInsertObject(RLMManagedSet *set, id object) {
+    RLMSetValidateMatchingObjectType(set, object);
+    changeSet(set, ^{
+        RLMAccessorContext context(*set->_objectInfo);
+        set->_backingSet.insert(context, object);
+    });
+}
+
+static void RLMRemoveObject(RLMManagedSet *set, id object) {
+    RLMSetValidateMatchingObjectType(set, object);
+    changeSet(set, ^{
+        RLMAccessorContext context(*set->_objectInfo);
+        set->_backingSet.remove(context, object);
+    });
+}
+
+static void ensureInWriteTransaction(NSString *message, RLMManagedSet *set, RLMManagedSet *otherSet) {
+    if (!set.realm.inWriteTransaction && !otherSet.realm.inWriteTransaction) {
+        @throw RLMException(@"Can only perform %@ in a Realm in a write transaction - call beginWriteTransaction on an RLMRealm instance first.", message);
+    }
+}
+
+- (void)addObject:(id)object {
+    RLMInsertObject(self, object);
+}
+
+- (void)addObjects:(id<NSFastEnumeration>)objects {
+    changeSet(self, ^{
+        RLMAccessorContext context(*_objectInfo);
+        for (id obj in objects) {
+            RLMSetValidateMatchingObjectType(self, obj);
+            _backingSet.insert(context, obj);
+        }
+    });
+}
+
+- (void)removeObject:(id)object {
+    RLMRemoveObject(self, object);
+}
+
+- (void)removeAllObjects {
+    changeSet(self, ^{
+        _backingSet.remove_all();
+    });
+}
+
+- (RLMManagedSet *)managedObjectFrom:(RLMSet *)set {
+    for (id obj in set) {
+        RLMSetValidateMatchingObjectType(self, obj);
+    }
+    if (!set.realm) {
+        @throw RLMException(@"Right hand side value must be a managed Set.");
+    }
+
+    if (_type != set.type) {
+        @throw RLMException(@"Set must match type of \"self\" '%@'", RLMTypeToString(_type));
+    }
+
+    return static_cast<RLMManagedSet *>(set);
+}
```

Comment:
```suggestion
- (RLMManagedSet *)managedObjectFrom:(RLMSet *)set {
    auto managedSet = RLMDynamicCast<RLMSet>(set);
    if (!managedSet) {
        @throw RLMException(@"Right hand side value must be a managed Set.");
    }
    if (_type != managedSet->_type || _objectInfo != managedSet->_objectInfo) {
        @throw RLMException(@"Set must match type of \"self\" '%@'", RLMTypeToString(_type));
    }
    return managedSet;
}
```

seems like a more correct implementation of this function, but not allowing an unmanaged set on the RHS is odd here.

Reply:
`managedObjectFrom` is used for methods that mutate and compare against the managed object, so I thought it was right to make a helper that could perform the cast.